File: 算法笔记/NQSDemo/src/main/java/jeason/A_Leetcode/其他零散题/_204_计算质数的个数.java
package jeason.A_Leetcode.其他零散题;


// https://leetcode-cn.com/problems/count-primes/
// 统计所有小于非负整数 n 的质数的数量。
// 输入：n = 10
// 输出：4
// 解释：小于 10 的质数一共有 4 个, 它们是 2, 3, 5, 7 。

public class _204_计算质数的个数 {
    //--------方法1:暴力方法----------
    public int countPrimes(int n) {
        int count = 0;
        for (int i = 2; i < n; ++i) {
            count += isPrime(i) ? 1 : 0;
        }
        return count;
    }
    public boolean isPrime(int x) {
        //这里使用i*i相当于对x开平方，减少一半重复的遍历
        for (int i = 2; i * i <= x; ++i) {
            if (x % i == 0) {
                return false;
            }
        }
        return true;
    }

    //--------方法2：埃氏筛选--------
    //思路:素数的量少，合数的量大，开始默认都为素数，把合数遍历标记完后，剩下的都是素数了
    public int countPrimes2(int n) {
        boolean[] isPrime = new boolean[n];//默认都是false标识素数
        int count = 0;
        for(int i=2; i<n; i++){

            if(!isPrime[i]){ //bool数组没有标记为素数，就往下进行
                count++;

                //写法1：每次i递增之后，都会从2开始到n-1去乘i。有重复遍历的部分
                for (int t=2*i; t<n; t+=i){
                    // 2*i+i 之后每次t+=i,相当于2*i 3*i 4*i 5*i...了
                    isPrime[t] = true;
                }

                //写法2：每次不再从2开始遍历去乘i，避免重复遍历
                for (int t=i*i; t<n; t+=i){
                    // 2*i+i 之后每次t+=i,相当于2*i 3*i 4*i 5*i...了
                    isPrime[t] = true;
                }
                
            }
        }
        return count;
    }

}

File: PythonAlgorithm/Leetcode/其他零散题/_204_计算质数的个数.py

# https://leetcode.cn/problems/count-primes/

""" 
给定整数 n ，返回 所有小于非负整数 n 的质数的数量 。

输入：n = 10
输出：4
解释：小于 10 的质数一共有 4 个, 它们是 2, 3, 5, 7 。

题解：https://leetcode.cn/problems/count-primes/solutions/36629/pythonzui-you-jie-fa-mei-you-zhi-yi-liao-ba-by-bru/
"""


class Solution(object):
    
    def countPrimes(self, n: int) -> int:
        # result = self.countPrimes1(n)
        result = self.countPrimes2(n)
        return result
    
    # --------方法1:暴力遍历方法----------
    def countPrimes1(self, n: int) -> int:
        if n < 2:
            return 0
        count: int = 0
        for i in range(2, n):
            isPrime: bool = True
            for j in range(2, i):
                if i % j == 0:
                    isPrime = False
                    break
            count += int(isPrime)
        return count
        
    # --------方法2：埃氏筛选--------
    # 思路:素数的量少，合数的量大，开始默认都为素数，把合数遍历标记完后，剩下的都是素数了
    def countPrimes2(self, n: int) -> int:
        if n < 2:
            return 0
        
        # 默认都是质数
        flagList: list = [1] * n
        
        # 0和1不是质数，先排除掉
        flagList[0] = flagList[1] = 0
        
        # 遍历范围：2到n的开根号，减少遍历次数
        for i in range(2, int(n ** 0.5)+1):
            if flagList[i]:
                flagList[i*i: n: i] = [0] * ((n - 1 - i * i) // i + 1)
        return sum(flagList)
            



def run():
    result: int = Solution().countPrimes(20)
    print(f"{result}")

run()

